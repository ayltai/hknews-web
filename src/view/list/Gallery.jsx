import React from 'react';
import { SingleFieldList } from 'react-admin';
import Lightbox from 'react-image-lightbox';
import 'react-image-lightbox/style.css';

export class Gallery extends React.PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            index  : 0,
            isOpen : false,
        };
    }

    getImages = () => {
        return Object.entries(this.getEntries()).map(record => ({
            caption : record[1][this.props.caption],
            source  : record[1][this.props.source],
        }));
    }

    getEntries = () => this.props.ids && this.props.data ? this.props.ids.map(id => this.props.data[id]) : [];

    render = () => {
        const { source, ...props } = this.props;
        const { index, isOpen }    = this.state;

        const entries = this.getEntries();
        const images  = this.getImages();

        const children = React.Children.map(this.props.children, child => {
            const element = React.cloneElement(child, {
                onClick : event => this.setState({
                    index  : Math.max(0, entries ? entries.findIndex(entry => entry[this.props.source] === event.target.src) : 0),
                    isOpen : true,
                }),
            });

            return element;
        });

        return (
            <div>
                <SingleFieldList
                    linkType={false}
                    {...props}>
                    {children[0]}
                </SingleFieldList>
                {isOpen && (
                    <Lightbox
                        mainSrc={images[index].source}
                        nextSrc={index + 1 < images.length ? images[(index + 1) % images.length].source : undefined}
                        prevSrc={index - 1 >= 0 ? images[(index + images.length - 1) % images.length].source : undefined}
                        imageCaption={images[index].caption}
                        onCloseRequest={() => this.setState({
                            isOpen : false,
                        })}
                        onMoveNextRequest={() => this.setState({
                            index : (index + 1) % images.length,
                        })}
                        onMovePrevRequest={() => this.setState({
                            index : (index + images.length - 1) % images.length,
                        })}
                    />
                )}
            </div>
        );
    }
}
