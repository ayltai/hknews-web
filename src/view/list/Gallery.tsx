import React from 'react';
import Lightbox from 'react-image-lightbox';
import 'react-image-lightbox/style.css';

const { SingleFieldList } = require('react-admin');

interface IProps {
    children : any;
    ids?     : string[];
    data?    : any[string];
    caption  : string;
    source   : string;
}

interface IState {
    index  : number;
    isOpen : boolean;
}

export class Gallery extends React.PureComponent<IProps, IState> {
    public constructor(props : IProps) {
        super(props);

        this.state = {
            index  : 0,
            isOpen : false,
        };
    }

    private getImages : () => any[] = () : any[] => {
        return Object.entries(this.getEntries()).map((record : any[]) => ({
            caption : record[1][this.props.caption],
            source  : record[1][this.props.source],
        }));
    }

    private getEntries : () => any = () : any => this.props.ids && this.props.data ? this.props.ids.map((id : string) => this.props.data[id]) : [];

    public render : () => React.ReactNode = () : React.ReactNode => {
        const { source, ...props } = this.props;
        const { index, isOpen }    = this.state;

        const entries : any   = this.getEntries();
        const images  : any[] = this.getImages();

        const children : any = React.Children.map(this.props.children, (child : any) => {
            const element : any = React.cloneElement(child, {
                onClick : (event : any) : void => this.setState({
                    index  : Math.max(0, entries ? entries.findIndex((entry : any) => entry[this.props.source] === event.target.src) : 0),
                    isOpen : true,
                }),
            });

            return element;
        });

        return (
            <div>
                <SingleFieldList
                    linkType={false}
                    {...props}>
                    {children[0]}
                </SingleFieldList>
                {isOpen && (
                    <Lightbox
                        mainSrc={images[index].source}
                        nextSrc={index + 1 < images.length ? images[(index + 1) % images.length].source : undefined}
                        prevSrc={index - 1 >= 0 ? images[(index + images.length - 1) % images.length].source : undefined}
                        imageCaption={images[index].caption}
                        onCloseRequest={() : void => this.setState({
                            isOpen : false,
                        })}
                        onMoveNextRequest={() : void => this.setState({
                            index : (index + 1) % images.length,
                        })}
                        onMovePrevRequest={() : void => this.setState({
                            index : (index + images.length - 1) % images.length,
                        })}
                    />
                )}
            </div>
        );
    }
}
