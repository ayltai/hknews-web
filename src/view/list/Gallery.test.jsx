import { mount, } from 'enzyme';
import React from 'react';
import { TestContext, } from 'react-admin';
import renderer from 'react-test-renderer';

import { Gallery, } from './Gallery';

describe('', () => {
    it('', () => {
        expect(renderer.create(
            <TestContext>
                <Gallery
                    ids={[ 'id1', 'id2' ]}
                    data={{
                        id1 : {
                            caption : '1',
                            source  : 'image1',
                        },
                        id2 : {
                            caption : '2',
                            source  : 'image2',
                        },
                    }}
                    caption='caption'
                    source='source'>
                    <img src='image1' />
                </Gallery>
            </TestContext>
        ).toJSON()).toMatchSnapshot();
    });

    it('mount without errors', () => {
        mount(
            <TestContext>
                <Gallery
                    ids={[ 'id1', 'id2' ]}
                    data={{
                        id1 : {
                            caption : '1',
                            source  : 'image1',
                        },
                        id2 : {
                            caption : '2',
                            source  : 'image2',
                        },
                    }}
                    caption='caption'
                    source='source'>
                    <img src='image1' />
                </Gallery>
            </TestContext>
        );
    });
});
