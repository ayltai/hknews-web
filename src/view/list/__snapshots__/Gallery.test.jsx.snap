// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div
    caption="caption"
    className="RaSingleFieldList-root-1"
  >
    <img
      onClick={[Function]}
      record={
        Object {
          "caption": "1",
          "source": "image1",
        }
      }
      src="image1"
    />
    <img
      onClick={[Function]}
      record={
        Object {
          "caption": "2",
          "source": "image2",
        }
      }
      src="image1"
    />
  </div>
</div>
`;
